### PR TITLE
Feature: Http 2 support for Server

### LINKED ISSUE #186: Support for Http2
This should be straightforward. Some documentation getting started material

https://www.baeldung.com/netty-http2

### PR DESCRIPTION
**Pull request by dinojohn**
http2 support for server

### CODE REVIEW

File: zio-http/src/main/scala/zhttp/service/Server.scala [hunk 1]
```diff
@@ -39,6 +46,7 @@ object Server {
     maxRequestSize: Int = 4 * 1024, // 4 kilo bytes
     error: Option[Throwable => ZIO[R, Nothing, Unit]] = None,
     sslOption: ServerSSLOptions = null,
+    enableHttp2: Boolean = false,
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/service/Server.scala:51]:
```suggestion
    http2: Boolean = false,
```

File: zio-http/src/main/scala/zhttp/service/Server.scala [hunk 2]
```diff
@@ -22,6 +28,7 @@ sealed trait Server[-R, +E] { self =>
     case MaxRequestSize(size) => s.copy(maxRequestSize = size)
     case Error(errorHandler)  => s.copy(error = Some(errorHandler))
     case Ssl(sslOption)       => s.copy(sslOption = sslOption)
+    case EnableHttp2(enable)  => s.copy(enableHttp2 = enable)
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/service/Server.scala]:
```suggestion
    case Http2  => s.copy(enableHttp2 = true)
```

File: zio-http/src/main/scala/zhttp/service/Server.scala [hunk 3]
```diff
@@ -48,12 +56,14 @@ object Server {
   private final case class App[R, E](http: HttpApp[R, E])                             extends Server[R, E]
   private final case class Error[R](errorHandler: Throwable => ZIO[R, Nothing, Unit]) extends Server[R, Nothing]
   private final case class Ssl(sslOptions: ServerSSLOptions)                          extends UServer
+  private final case class EnableHttp2(enable: Boolean)                               extends UServer
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/service/Server.scala]:
```suggestion
  private final case object Http2                               extends UServer
```

File: zio-http/src/main/scala/zhttp/service/Server.scala [hunk 4]
```diff
@@ -48,12 +56,14 @@ object Server {
...
 
   def app[R, E](http: HttpApp[R, E]): Server[R, E]                                   = Server.App(http)
   def maxRequestSize(size: Int): UServer                                             = Server.MaxRequestSize(size)
   def port(int: Int): UServer                                                        = Server.Port(int)
   def error[R](errorHandler: Throwable => ZIO[R, Nothing, Unit]): Server[R, Nothing] = Server.Error(errorHandler)
   def ssl(sslOptions: ServerSSLOptions): UServer                                     = Server.Ssl(sslOptions)
+  def enableHttp2(enable: Boolean): UServer                                          = Server.EnableHttp2(enable)
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/service/Server.scala]:
```suggestion
  def http2: UServer                                          = Server.EnableHttp2(enable)
```

File: zio-http/src/main/scala/zhttp/service/DecodeJRequest.scala [hunk 5]
```diff
@@ -15,5 +16,9 @@ trait DecodeJRequest {
     endpoint = method -> url
     data     = HttpData.fromByteBuf(jReq.content())
   } yield Request(endpoint, headers, data, ctx)
-
+  def decodeHttp2Header(hh: JHttp2HeadersFrame): Either[HttpError, Request]                           = for {
+    url <- URL.fromString(hh.headers().path().toString)
+    method   = Method.fromString(hh.headers().method().toString)
+    endpoint = method -> url
+  } yield Request(endpoint)
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/service/DecodeJRequest.scala]:
```suggestion
  } yield Request(endpoint, headers, data, ctx)
```

File: zio-http/src/main/scala/zhttp/service/EncodeResponse.scala [hunk 6]
```diff
@@ -36,4 +37,20 @@ trait EncodeResponse {
     }
     new JDefaultHttpResponse(jVersion, jStatus, jHttpHeaders)
   }
+
+  def encodeResponse[R, E](res: Response.HttpResponse[R, E]): JHttp2Headers = {
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/service/EncodeResponse.scala:41]: add unit test for encoding only.

File: zio-http/src/main/scala/zhttp/service/server/ServerChannelInitializer.scala [hunk 7]
```diff
@@ -1,41 +1,95 @@
...
+        .addFirst(SSL_HANDLER, new OptionalSSLHandler(httpH, http2H, sslctx, settings))
       ()
-    }
-    channel
-      .pipeline()
+    } else configureClearText(httpH, http2H, channel, settings)
+  }
+}
+
+object ServerChannelInitializer {
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/service/server/ServerChannelInitializer.scala]: Try creating separate channel initializers for Http, Http2, SSL etc.

File: zio-http/src/main/scala/zhttp/http/Header.scala [hunk 8]
```diff
@@ -28,6 +29,13 @@ object Header {
       .map(h => Header(h.getKey, h.getValue))
       .toList
 
+  def make(headers: Http2Headers): List[Header] =
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/http/Header.scala:32]:
```suggestion
  def fromHttp2Headers(headers: Http2Headers): List[Header] =
```

**Comment by dinojohn** [comment]
closing in favour of #332

**Pull request closed without merging**